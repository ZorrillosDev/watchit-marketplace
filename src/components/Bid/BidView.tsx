// REACT IMPORTS
import React, { FC, useState } from 'react'

// MUI IMPORTS
import { Box, BoxProps, Button, Grid, styled, TextField, Theme, Typography } from '@mui/material'
import { LoadingButton } from '@mui/lab'

// PROJECT IMPORTS
import Modal from '@components/Modal'
import { Translation } from '@src/i18n'
import { SxProps } from '@mui/system'
import BidSummary from '@components/Bid/BidSummary'
import { useBalance } from '@hooks/useBalance'

// ===========================|| BID - VIEW ||=========================== //

export interface ModalBidViewProps {
  isLoading: boolean
  handleSetBid: (bid: number) => void
  buttonSx?: SxProps<Theme>
}

const BidView: FC<ModalBidViewProps> = (props): JSX.Element => {
  const [bidAmount, setBidAmount] = useState(0)
  const balance: number = useBalance()
  const [isOpen, setOpen] = useState(false)
  const onClose = (): void => setOpen(false)
  const onOpen = (): void => setOpen(true)

  return (
    <>
      <Button variant='contained' color='primary' size='large' onClick={() => onOpen()} sx={props.buttonSx}>
        <Translation target='MOVIE_PROFILE_PRICE_MAKE_OFFER' />
      </Button>

      <Modal
        isOpen={isOpen}
        onClose={() => onClose()}
      >
        <ModalBidContent>
          <Grid container spacing={2}>
            <Grid item xs={12}>
              <header>
                <Typography variant='h3'>
                  Place a bid
                </Typography>
              </header>
            </Grid>
            <Grid item xs={12}>
              <TextField
                id='bidAmount' name='bidAmount' label='Your bid' variant='outlined' fullWidth
                onChange={(e) => setBidAmount(+e.target.value)}
              />
            </Grid>
            <Grid item xs={12}>
              <BidSummary balance={balance} amount={bidAmount} fee={0.03} />
            </Grid>
            <Grid item xs={12}>
              <LoadingButton
                variant='contained'
                color='primary'
                size='large'
                loading={props.isLoading}
                onClick={() => props.handleSetBid(bidAmount)}
                fullWidth
              >
                Place a bid
              </LoadingButton>
            </Grid>
          </Grid>
        </ModalBidContent>
      </Modal>
    </>
  )
}

export default BidView

const ModalBidContent = styled(Box)<BoxProps>(({ theme }) => ({
  width: '100%',
  maxWidth: '22rem',
  padding: '1.5rem 1.5rem 1.5rem 2rem',
  [theme.breakpoints.down('sm')]: {
    maxWidth: '100%'
  }
}))
