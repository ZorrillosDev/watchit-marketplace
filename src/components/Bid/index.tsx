// ===========================|| BID EXPORTS ||=========================== //

export { BidContainer as default } from '@components/Bid/BidContainer'
