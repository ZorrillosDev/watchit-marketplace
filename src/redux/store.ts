import { createStore, applyMiddleware, compose, Store } from 'redux'
import { routerMiddleware } from 'connected-react-router/immutable'
import thunk from 'redux-thunk'
import rootReducer from './reducers'
import { History } from 'history'

export default (history: History, initialState = {}): Store => {
  // devtools for debugging in dev environment.
  // @ts-expect-error
  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ?? compose

  // Array of all middlewares to be applied.
  const middlewares = [
    thunk,
    routerMiddleware(history)
  ]

  const enhancers = [
    applyMiddleware(...middlewares)
  ]

  return createStore(
    rootReducer(history),
    initialState,
    composeEnhancers(...enhancers)
  )
}
