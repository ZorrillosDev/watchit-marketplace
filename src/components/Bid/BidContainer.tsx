// REACT IMPORTS
<<<<<<< HEAD
import React, { FC, useCallback, useEffect, useState } from 'react'
=======
import React, { FC, useCallback, useState } from 'react'
>>>>>>> main

// PROJECT IMPORTS
import BidView from '@components/Bid/BidView'
import { SxProps } from '@mui/system'
import { Theme } from '@mui/material'
import { MovieArgs, MoviesActions } from '@state/movies/types'
import { commitBidMovie } from '@state/movies/actions'
import { connect, RootStateOrAny } from 'react-redux'
import { useEthers } from '@usedapp/core'
import { useParams } from 'react-router'
import { MoviesResultState } from '@state/movies/reducer'

// ===========================|| BID - CONTAINER ||=========================== //

type BidContainerProps = { buttonSx?: SxProps<Theme> } & MoviesActions & MoviesResultState

const BidContainer: FC<BidContainerProps> = (props): JSX.Element => {
  const [isLoading, setIsLoading] = useState(false)
  const { id } = useParams<MovieArgs>()
  const { account } = useEthers()
  const {
    buttonSx,
    commitBidMovie
  } = props

  useEffect(() => {
    if (Object.is(props.response, undefined)) {
      return
    }
    setIsLoading(false)
  }, [props.response])

  const handleSetBid = useCallback((bid: number): void => {
    if (account === undefined) {
      return
    }

    setIsLoading(true)
    commitBidMovie({
      account: account ?? '',
      id,
      bid
    })
  }, [account])

  const clickHandler = (): void => {
    setIsLoading(true)
  }

  return <BidView {...{ buttonSx, isLoading, handleSetBid, clickHandler }} />
}

const mapDispatchToProps: Partial<MoviesActions> = { commitBidMovie }
const mapStateToProps = (state: RootStateOrAny): MoviesResultState => {
  return {
    response: state.movies.response
  }
}

export const Bid = connect(
  mapStateToProps,
  mapDispatchToProps
)(BidContainer)
