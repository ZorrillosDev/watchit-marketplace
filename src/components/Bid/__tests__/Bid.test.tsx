import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import { store } from '@state/store';
import Bid from '@components/Bid';

/* eslint-disable no-undef */

jest.mock('react-router', () => ({
  useParams: () => {
    return { id: '123' };
  },
}));

describe('<Bid />', () => {
  it('should render', () => {
    const component = mount(
      <Provider store={store}>
        <Bid />
      </Provider>,
    );

    expect(component).toMatchSnapshot();
  });
});
