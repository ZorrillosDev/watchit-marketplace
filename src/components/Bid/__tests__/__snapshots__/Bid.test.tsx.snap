// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Bid /> should render 1`] = `ReactWrapper {}`;
